Follow Commenters

justdvl opened the pull request:
New method added, to follow all users who commented on speific user(s).
As requested here: https://github.com/timgrossmann/InstaPy/issues/1659#issuecomment-374047686
To run, add to quickstart.py, this:

session.follow_commenters(['mody._.21', 'samofficiel', 'jesus2018m']) 

It works most of the times perfectly but crash on rare occasions, I'd be happy to get help with that.

Tjorriemorrie commented on instapy/follow_commenters.py:
```diff
@@ -0,0 +1,186 @@
+"""Methods to extract the data for the given usernames profile"""
```
all the files ends with *_util, do you mind changing this to `commenters_util.py`?

Tjorriemorrie commented on instapy/follow_commenters.py:
```diff
@@ -0,0 +1,186 @@
+def get_user_info(browser):
+  """Get the basic user info from the profile screen"""
+
+  container = browser.find_element_by_class_name('_mesn5')
```
You cannot use css class names. It's generated and will break everytime IG make a release.

justdvl replied:
how to replace:
    post = browser.find_element_by_class_name('_622au') ?
because:    
    post = browser.find_element_by_xpath("/html/body/div[4]/div/div[2]/div/article")
doesnt work.

Tjorriemorrie commented on instapy/follow_commenters.py:
```diff
@@ -0,0 +1,186 @@
+  print ("\nbio: ", bio,"\n")
+  prof_img = img_container.find_element_by_tag_name('img').get_attribute('src')
+  num_of_posts = int(infos[0].text.split(' ')[0].replace(',', ''))
+  followers = infos[1].text.split(' ')[0].replace(',', '').replace('.', '')
```
this has been done. refer `format_number` from util.py

Tjorriemorrie commented on instapy/follow_commenters.py:
```diff
@@ -0,0 +1,186 @@
+from selenium.webdriver.common.keys import Keys
+
+def get_user_info(browser):
+  """Get the basic user info from the profile screen"""
```
4 spaces. not 2. follow pep8

Tjorriemorrie commented on instapy/follow_commenters.py:
```diff
@@ -0,0 +1,186 @@
+
+  post = browser.find_element_by_class_name('_622au')
+
+  #print('BEFORE IMG')
```
remove

Tjorriemorrie commented on instapy/follow_commenters.py:
```diff
@@ -0,0 +1,186 @@
+  # or the button to display all the comments
+  comments = []
+  tags = []
+  
```
max 1 line space inside function. follow pep8

Tjorriemorrie commented on instapy/instapy.py:
```diff
@@ -468,6 +473,19 @@ def clarifai_check_img_for(self, tags=None, comment=False, comments=None):
 
         return self
 
+    def follow_commenters(self, nicknamegetarr):
```
what is getarr? Do you mind renaming this please? `nicknames` should be fine?

Tjorriemorrie commented on instapy/instapy.py:
```diff
@@ -468,6 +473,19 @@ def clarifai_check_img_for(self, tags=None, comment=False, comments=None):
         return self
 
+    def follow_commenters(self, nicknamegetarr):
+        print ("running fc")
```
please do not use abbreviations.

Tjorriemorrie commented on instapy/instapy.py:
```diff
@@ -468,6 +473,19 @@ def clarifai_check_img_for(self, tags=None, comment=False, comments=None):
+    def follow_commenters(self, nicknamegetarr):
+        print ("running fc")
+        for nicknameget in nicknamegetarr:
+          self.nicknameget = nicknameget
```
Please do not assign loop variables to the instance.

Tjorriemorrie commented on instapy/instapy.py:
```diff
@@ -468,6 +473,19 @@ def clarifai_check_img_for(self, tags=None, comment=False, comments=None):
 
+    def follow_commenters(self, nicknamegetarr):
+        print ("running fc")
+        for nicknameget in nicknamegetarr:
```
just `nickname` should be fine.

Tjorriemorrie commented on instapy/instapy.py:
```diff
@@ -468,6 +473,19 @@ def clarifai_check_img_for(self, tags=None, comment=False, comments=None):
+          self.nicknameget = nicknameget
+          user_commented_list = extract_information(self.browser, self.nicknameget)
+          print (user_commented_list)
+          if (len(user_commented_list))>0:
```
space around ` > `. If you are using pycharm they have a Reformat code utility.

Tjorriemorrie commented on instapy/instapy.py:
```diff
@@ -509,6 +527,8 @@ def set_lower_follower_count(self, limit=None):
         self.like_by_followers_lower_limit = limit or 0
         return self
 
+
```
undo: only 1 space line between functions in a class. pep8 is your friend.

justdvl commented:
Thank you Tjorriemorrie for the comments, I am working on fixing these issues.

I found out my script was crashing on some occasions, because of self aborting line in:

    def follow_by_list(self, followlist, times=1): 
        """Allows to follow by any scrapped list"""
        print ("started")
        self.follow_times = times or 0
        if self.aborting:
            return self

This return self should be printing something to terminal, so we know why the script suddenlty ends.
If I uncomment return self, all works great.
Can somone explain to me what is the function of self aborting? (I am new to this project)

Thanks

justdvl commented:
All suggestions fixed.

Added customizations:

Call by:

 session.follow_commenters(array, amount, daysold, max_pic)

array = array of usernames, to scan
amount = max amount of people to follow
daysold = maximum age of pictures  to analyze
max_pic = maximum number of pictures to analyze

CharlesCCC commented on instapy/commenters_util.py:
```diff
@@ -0,0 +1,219 @@
+from .util import get_number_of_posts
+
+#remove duplicates from arrawy while preserving order
+def f7(seq):
```
what `f7` stand for ?

justdvl replied:
Its a function that:
#remove duplicates from array while preserving order

It is because I am getting users who commented on number of posts, and I decided to sort them by frequency - to have users who comments the most be the first ones to follow; but at the same time removing duplicates.

CharlesCCC replied:
@justdvl yea, I can read the comment and I can read that part of code to. I'm just questioning the function naming itself  why named this method `f7`,  like why not `f8` `f9`... ??? For code readability point, I think we should only need left comment if it really necessary and there is no way to find a good method name to explain itself.

justdvl replied:
Can be changed, it's minor function copied from internet. Name was given to it because of speed..
I tried to comment a lot because the code seems to be bit too long..

justdvl replied:
You are right. I have renammed the function.

justdvl commented:
Added function follow_likers, which will get the _amount_ of users who liked specific photo by _url_
Next functions can be build on top of this.

Use:

    session.follow_likers("https://www.instagram.com/p/BgO7uCihPl8/", amount = 50)

where amount is how many users to follow (and how many likes to scrap)

as requested: https://github.com/timgrossmann/InstaPy/issues/1680

justdvl commented:
I am also thinking, why are we following users by going to each profile one by one, if we could be just clicking Follow buttons when we have dialog with user's list opened.
It is also more natural way, coming from Instagram bot detection place.

justdvl commented:
Added module: Follow_user_likers  … example to put into quickstart.py:

` session.follow_user_likers (['natgeotravel' , 'jesus2018m'], photos_grab_amount = 2, follow_likers_per_photo = 2, randomize=True)
`

timgrossmann commented:
@justdvl Please make sure to also include explanation in the README file.

Tjorriemorrie commented on instapy/instapy.py:
```diff
@@ -473,9 +473,10 @@ def clarifai_check_img_for(self, tags=None, comment=False, comments=None):
-    def follow_likers(self, photo_url, amount=2):
-        user_liked_list = users_liked (self.browser, photo_url, amount)
-        self.follow_by_list(user_liked_list[:amount])
+    def follow_likers(self, photo_url_arr, amount=10):
```
Please don't use Hungarian notation, `photo_urls` is more pythonic.
FYI: you can also add supporting a string by:
```
if not isinstance(photo_urls, list):
    photo_urls = [photo_urls]
```

CharlesCCC commented on instapy/commenters_util.py:
```diff
@@ -0,0 +1,340 @@
+def extract_information(browser, nick, daysold, max_pic):
+  
+    """Get all the information for the given username"""
+    browser.get('https://www.instagram.com/' + nick)
```
In here you used `nick` for username ?  Can we keep it consistent to use `username` as well, so we don't confuse others.  thank you.

justdvl commented:
Sorry, I dont underatand, didnt I just do it? I replaced nick for username.
My laptop battery just run oit so Ill be able to do it later at evening.
I will also split long code into more smaller functions to make it easier to read and easier to customize.

CharlesCCC commented:
@justdvl Yea I saw you did, thank you for the change. 
maybe @timgrossmann looked at the wrong commit  😕 ?

timgrossmann commented on instapy/instapy.py:
```diff
@@ -467,19 +473,55 @@ def clarifai_check_img_for(self, tags=None, comment=False, comments=None):
+            self.follow_by_list(user_liked_list[:amount])
+        return self
+
+    def follow_commenters(self, nick_array, amount=100, daysold=365, max_pic = 100):
```
I just saw this... Sorry for not pointing out what I saw right away

timgrossmann commented on README.md:
```diff
@@ -238,6 +253,43 @@ session.follow_user_followers(['friend1', 'friend2', 'friend3'], amount=10, rand
+# The amount is how many people to follow
+# In this case 10 people who liked photo in photo_url_arr will be followed, for each photo
+
+def follow_likers(self, photo_url_arr, amount=10):
```
Could you please also paste the actual way to call this method in the quickstart here?

timgrossmann commented:
@justdvl Sorry for the many changes and resolves...
Just one more thing then we can finally add this to InstaPy!

Thank you very much for the hard work

gaikaz commented:
Hi. I'm just wondering, why is this still not pulled to master?

timgrossmann commented:
@gaikaz Because no-one has resolved the merge conflicts, yet...

argentgagner commented:
It doesn't seem to be working... See #2025

justdvl commented:
I have fixed the changes instagram did

https://github.com/timgrossmann/InstaPy/pull/2134
